library/std/src/sys/windows/process.rs

```diff
@@ -31,13 +32,66 @@ use libc::{c_void, EXIT_FAILURE, EXIT_SUCCESS};
 // Command
 ////////////////////////////////////////////////////////////////////////////////
 
-#[derive(Clone, Debug, Eq, PartialEq, Ord, PartialOrd)]
+#[derive(Clone, Debug, Eq)]
 #[doc(hidden)]
 pub struct EnvKey(OsString);
 
+// Windows environment variables are case preserving but comparisions are case
+// insensitive. So we need to convert to uppercase. This converts UTF-16 to
+// UTF-32 code units, converting the case along the way. Unpaired surrogates are
+// left as they are.
+fn iter_uppercase<T: Iterator<Item = u16>>(iter: T) -> impl Iterator<Item = u32> {
+    char::decode_utf16(iter).flat_map(|cp| match cp {
+        Err(e) => CodeUnits::Unit(Some(e.unpaired_surrogate())),
+        Ok(cp) => CodeUnits::Iter(cp.to_uppercase()),
```

Comment:
I checked and `to_uppercase` is not always correct: `'ß'.to_uppercase() == "SS"`, but the following fails
```rust
set_var("ß", "something");
var("SS").unwrap() // panic: Err<NotPresent>
```

I can't find anything about in official documentation, but from tests I did I suspect Windows uses simple case folding (for context see [Unicode Technical Report #21 Case Mappings Section 1.3](https://www.unicode.org/reports/tr21/tr21-5.html), simple case folding is the unicode case folding algorithm, but restricted to code points that don't expand under upper/lower casing like `ß`). I also checked if any normalization or locale-dependent behavior was happening, but that doesn't appear to be the case: `é` and `é` (e + acute accent combining character) are different keys and nothing changed around `I` on the Turkish locale (https://en.wikipedia.org/wiki/Dotted_and_dotless_I).

Not sure what the best way to use case folding in std would be (if that is even what Windows is indeed using). There are crates like [unicode-rs/rust-caseless](https://github.com/unicode-rs/rust-caseless) implementing case folding, but I couldn't find any that also implements the simple variant.

Reply:
Hm yeah, that is a problem. I think the only way to resolve this properly is to use the OS's own function for this. So in the next commit I've replaced my implementation with a Windows API call and expanded the testing slightly.

Reply:
I looked through the docs some more: 

File paths and environment variables should be compared using `StringComparison.OrdinalIgnoreCase` [1]. Comparisons made using `OrdinalIgnoreCase` are behaviorally the composition of two calls: calling `ToUpperInvariant` on both string arguments, and doing a comparison using `Ordinal` [2]. A comparison using `Ordinal` is just a comparison based on the numerical value of each utf16 code unit [3]. 

So the original approach was conceptually correct, but Windows apparently handles converting to uppercase differently than rust (I haven't yet found anywhere how `ToUpperInvariant` is actually implemented 🙁). I agree with the approach to instead just use the Windows API. 

(As an aside, with the relevant Windows API added a future possibility for `std` would be to expose case insensitive functionality with a `std::os` extension trait, for example for correctly comparing paths case insensitively.)

[1] https://docs.microsoft.com/en-us/dotnet/standard/base-types/best-practices-strings#choosing-a-stringcomparison-member-for-your-method-call
[2] https://docs.microsoft.com/en-us/dotnet/standard/base-types/best-practices-strings#stringtoupper-and-stringtolower
[3] https://docs.microsoft.com/en-us/dotnet/api/system.stringcomparison?view=net-5.0#System_StringComparison_Ordinal

Reply:
Can you add this conclusion, why CompareStringOrdinal is used, in a comment in the code?

Reply:
@m-ou-se I've [added a comment](https://github.com/rust-lang/rust/pull/85270/commits/a200c01e4fb32179af1cbbc47a1e5ea7a6394e9a) (based on CDirkx's) to briefly explain how comparing environment keys works and thus why `CompareStringOrdinal` is used.